### PR TITLE
Add method to get the federated user id

### PR DESCRIPTION
**Pull request by ThaminduR**
### Proposed changes in this pull request

Add method to get federated user id from `username`, `tenantId`, `IdpId`.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 1]
```diff
@@ -871,4 +872,19 @@ public int getActiveSessionCount(String tenantDomain) throws UserSessionExceptio
...
+    /**
+     * Returns the user id of the federated user.
+     *
+     * @param userName - username of the federated user.
+     * @param tenantId
+     * @param idPId
+     * @return
+     * @throws UserSessionException
+     */
+    public String getFederatedUserId(String userName, int tenantId, int idPId)
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:885]:
```suggestion
    public String getFederatedUserId(String subjectIdentifier, int tenantId, int idPId)
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 2]
```diff
@@ -871,4 +872,19 @@ public int getActiveSessionCount(String tenantDomain) throws UserSessionExceptio
         }
         return activeSessionCount;
     }
+
+    /**
+     * Returns the user id of the federated user.
+     *
+     * @param userName - username of the federated user.
+     * @param tenantId
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:880]: tenantId of?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 3]
```diff
@@ -871,4 +872,19 @@ public int getActiveSessionCount(String tenantDomain) throws UserSessionExceptio
...
         return activeSessionCount;
     }
+
+    /**
+     * Returns the user id of the federated user.
+     *
+     * @param subjectIdentifier Subject Identifier of the federated user.
+     * @param tenantId          Id of the tenant domain
+     * @param idPId
+     * @return
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:882]: improve the comments

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 4]
```diff
@@ -871,4 +872,21 @@ public int getActiveSessionCount(String tenantDomain) throws UserSessionExceptio
...
+    /**
+     * Returns the user id of the federated user.
+     *
+     * @param subjectIdentifier - Subject Identifier of the federated user.
+     * @param tenantId          - Id of the tenant domain.
+     * @param idPId             - Id of the identity provider.
+     * @return userId - User Id of the federated user.
+     * @throws UserSessionException
+     */
+    public String getFederatedUserId(String subjectIdentifier, int tenantId, int idPId)
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:885]: What would be the tenantId here?
Is it the IDP tenant ID?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 5]
```diff
@@ -871,4 +872,21 @@ public int getActiveSessionCount(String tenantDomain) throws UserSessionExceptio
...
+            throws UserSessionException {
+
+        //When federated user is stored, the userDomain is passed as null and it is added as "FEDERATED" to the store.
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:888]: improve the comment formatting. There should be a space after //

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 6]
```diff
@@ -871,4 +872,21 @@ public int getActiveSessionCount(String tenantDomain) throws UserSessionExceptio
...
+        //Here null is passed as userDomain to get the userId from the "FEDERATED" userDomain.
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:889]: The above comment also has a similar meaning. This is redundant

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 7]
```diff
@@ -871,4 +872,21 @@ public int getActiveSessionCount(String tenantDomain) throws UserSessionExceptio
         }
         return activeSessionCount;
     }
+
+    /**
+     * Returns the user id of the federated user.
+     *
+     * @param subjectIdentifier - Subject Identifier of the federated user.
+     * @param tenantId          - Id of the tenant domain.
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:880]: mention which tenant is this. whether IDP tenant/ sp tenant

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java (hunk 4 above)
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:885]: [Here](https://github.com/wso2/carbon-identity-framework/blob/master/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java#L517 ) we get the tenant domain from the request and set it into the context and [here](https://github.com/wso2/carbon-identity-framework/blob/master/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractApplicationAuthenticator.java#L226) we set the tenant domain to the user from the context. From [here](https://github.com/wso2/carbon-identity-framework/blob/master/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java#L638) we get the tenant domain from the user and get the tenant id for that and [here](https://github.com/wso2/carbon-identity-framework/blob/master/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java#L656) we store the federated user into the user store.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 8]
```diff
@@ -871,4 +872,21 @@ public int getActiveSessionCount(String tenantDomain) throws UserSessionExceptio
...
+     * @param tenantId          - Id of the service provider's tenant domain.
+     * @param idPId             - Id of the identity provider.
+     * @return userId - User Id of the federated user.
+     * @throws UserSessionException
+     */
+    public String getFederatedUserId(String subjectIdentifier, int tenantId, int idPId)
+            throws UserSessionException {
+
+        // When federated user is stored, the userDomain is passed as null and it is added as "FEDERATED" to the store.
+        // Here null is passed as userDomain to get the userId from the "FEDERATED" userDomain.
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:889]:
```suggestion
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 9]
```diff
@@ -871,4 +872,21 @@ public int getActiveSessionCount(String tenantDomain) throws UserSessionExceptio
...
+        return getUserId(subjectIdentifier, tenantId, null, idPId);
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:890]:
```suggestion
        return getUserId(subjectIdentifier, tenantId, FEDERATED_USER_DOMAIN, idPId);
```

**Review by mefarazath — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 10]
```diff
@@ -34,6 +34,7 @@
 import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
 import org.wso2.carbon.identity.core.util.IdentityUtil;
 import org.wso2.carbon.idp.mgt.util.IdPManagementUtil;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:37]: Do we need this import?

Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java]: removed in 8adf6b5

**Pull request merged**